src/libprojectM/MilkdropPreset/BlurTexture.cpp

```diff
@@ -0,0 +1,363 @@
+#include "BlurTexture.hpp"
+
+#include "PerFrameContext.hpp"
+#include "PresetState.hpp"
+
+#include "MilkdropStaticShaders.hpp"
+
+#include <array>
+
+BlurTexture::BlurTexture()
+    : m_blurSampler(std::make_shared<Sampler>(GL_CLAMP_TO_EDGE, GL_LINEAR))
+{
+    auto staticShaders = libprojectM::MilkdropPreset::MilkdropStaticShaders::Get();
+
+    // Compile shader sources
+    m_blur1Shader.CompileProgram(staticShaders->GetBlurVertexShader(),
+                                 staticShaders->GetBlur1FragmentShader());
+    m_blur2Shader.CompileProgram(staticShaders->GetBlurVertexShader(),
+                                 staticShaders->GetBlur2FragmentShader());
+
+    m_blurFramebuffer.CreateColorAttachment(0, 0);
+
+    // Initialize Blur VAO/VBO with a single fullscreen quad.
+    static constexpr std::array<float, 16> pointsBlur{
+        -1.0, -1.0, 0.0, 0.0,
+        1.0, -1.0, 1.0, 0.0,
+        -1.0, 1.0, 0.0, 1.0,
+        1.0, 1.0, 1.0, 1.0};
```

Comment:
I had ChatGPT explain this to me

> This Vertex Buffer Object (VBO) represents a square (technically, two triangles) that cover the entire screen in Normalized Device Coordinates (NDC), with texture coordinates associated with each vertex.
> 
> Here's a breakdown of the values:
> 
> The first two values in each row (-1.0, -1.0, 1.0, -1.0, -1.0, 1.0, 1.0, 1.0) represent the x and y coordinates of four vertices in a 2D space. This is a square that covers the entire viewport in NDC, where (-1, -1) is the bottom left corner and (1, 1) is the top right corner.
> The third and fourth values in each row (0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 1.0, 1.0) represent the texture coordinates (u and v) associated with each vertex. They range from (0, 0) at the bottom left to (1, 1) at the top right of the texture.
> The drawing order of the vertices forms two triangles that together cover the entire viewport. The first triangle is formed by the vertices at (-1,-1), (1,-1), and (-1,1), and the second triangle is formed by the vertices at (-1,1), (1,-1), and (1,1).

Reply:
That's a very sophisticated explanation of "what is a quad?" 😅

Yeah, the blur textures are rendered as a simple quad, drawn with a specific gaussian blur shader. The quad also doesn't always fill the whole screen (or framebuffer, in this case), only in the first iteration. Subsequent blur steps will use only half or a quarter of the texture to create even more blurry images.

Reply:
yeah I just didn't get the texture coords part of it